src/approaches/interactive_learning_approach.py

```diff
@@ -0,0 +1,178 @@
+"""An approach that learns predicates from a teacher.
+"""
+
+import itertools
+from typing import Set, Callable, List, Sequence, Collection
+import numpy as np
+import torch
+from gym.spaces import Box
+
+from predicators.src import utils
+from predicators.src.approaches import TAMPApproach
+from predicators.src.structs import State, Predicate, ParameterizedOption, \
+    Type, Task, Action, Operator, Dataset, GroundAtom, ActionTrajectory, \
+    Object
+from predicators.src.operator_learning import learn_operators_from_data
+from predicators.src.models import MLPClassifier
+from predicators.src.utils import get_object_combinations
+from predicators.src.settings import CFG, get_save_path
+
+
+class InteractiveLearningApproach(TAMPApproach):
+    """An approach that learns predicates from a teacher.
+    """
+    def __init__(self,
+                 simulator: Callable[[State, Action], State],
+                 true_predicates: Set[Predicate],
+                 initial_options: Set[ParameterizedOption],
+                 types: Set[Type],
+                 action_space: Box,
+                 train_tasks: List[Task]) -> None:
+        # Only the teacher is allowed to know about the initial predicates
+        self._known_predicates = {p for p in true_predicates
+                                  if p.name in CFG.interactive_known_predicates}
+        predicates_to_learn = true_predicates - self._known_predicates
+        self._teacher = _Teacher(true_predicates, predicates_to_learn)
+        # All seen data
+        self.dataset: List[ActionTrajectory] = []
+        self.ground_atom_dataset: List[List[Set[GroundAtom]]] = []
+        # No cheating!
+        self._predicates = {_strip_predicate(p) for p in predicates_to_learn}
+        del true_predicates
+        del predicates_to_learn
+        super().__init__(simulator, self._predicates, initial_options,
+                         types, action_space, train_tasks)
+        self._operators: Set[Operator] = set()
+
+    @property
+    def is_learning_based(self) -> bool:
+        return True
+
+    def _get_current_operators(self) -> Set[Operator]:
+        assert self._operators, "Operators not learned"
+        return self._operators
+
+    def _get_current_predicates(self) -> Set[Predicate]:
+        return self._predicates
+
+    def learn_from_offline_dataset(self, dataset: Dataset) -> None:
+        # Get data from teacher
+        ground_atom_data = self._teacher.generate_data(dataset)
+        self.dataset.extend(dataset)
+        self.ground_atom_dataset.extend(ground_atom_data)
+        # Learn predicates
+        for pred in self._predicates:
+            print("\nPredicate:", pred)
+            assert pred not in self._known_predicates
+            positive_examples = []
+            negative_examples = []
+            # Positive examples
+            for i, trajectory in enumerate(self.ground_atom_dataset):
```

Comment:
It would be possible to reorganize this part of the code so that you only have to go through the dataset once total; right now it'll be 2*|predicates|. But the way that you have it written now is clear, so let's leave it this way for now. If in the future, datasets become so large that this part is slow, we can reorganize then.